src/main/java/io/odpf/firehose/sink/mongodb/MongoSink.java

```diff
@@ -0,0 +1,103 @@
+package io.odpf.firehose.sink.mongodb;
+
+import com.mongodb.BulkWriteError;
+import com.mongodb.MongoClient;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.WriteModel;
+import io.odpf.firehose.consumer.Message;
+import io.odpf.firehose.metrics.Instrumentation;
+import io.odpf.firehose.sink.AbstractSink;
+import io.odpf.firehose.sink.mongodb.request.MongoRequestHandler;
+import io.odpf.firehose.sink.mongodb.util.MongoResponseHandler;
+import org.bson.Document;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static io.odpf.firehose.metrics.Metrics.SINK_MESSAGES_DROP_TOTAL;
+
+/**
+ * MongoDB sink.
+ */
+public class MongoSink extends AbstractSink {
+
+    private final MongoRequestHandler mongoRequestHandler;
+    private final List<WriteModel<Document>> bulkRequest = new ArrayList<>();
+    private final List<String> mongoRetryStatusCodeBlacklist;
+    private final MongoCollection<Document> mongoCollection;
+    private final MongoClient mongoClient;
+    private List<Message> messages;
+
+    /**
+     * Instantiates a new Mongo sink.
+     *
+     * @param instrumentation               the instrumentation
+     * @param sinkType                      the sink type
+     * @param mongoCollection               the mongo collection
+     * @param mongoClient                   the mongo client
+     * @param mongoRequestHandler           the mongo request handler
+     * @param mongoRetryStatusCodeBlacklist the mongo retry status code blacklist
+     */
+    public MongoSink(Instrumentation instrumentation, String sinkType, MongoCollection<Document> mongoCollection, MongoClient mongoClient, MongoRequestHandler mongoRequestHandler,
+                     List<String> mongoRetryStatusCodeBlacklist) {
+        super(instrumentation, sinkType);
+        this.mongoCollection = mongoCollection;
+        this.mongoRequestHandler = mongoRequestHandler;
+        this.mongoClient = mongoClient;
+        this.mongoRetryStatusCodeBlacklist = mongoRetryStatusCodeBlacklist;
+    }
+
+    @Override
+    protected void prepare(List<Message> messageList) {
+        this.messages = messageList;
+        bulkRequest.clear();
+
+        messages.forEach(message -> bulkRequest.add(mongoRequestHandler.getRequest(message)));
+    }
+
+    @Override
+    protected List<Message> execute() throws Exception {
+        List<BulkWriteError> bulkWriteErrors = processRequest(bulkRequest);
+
+        if (bulkWriteErrors.isEmpty()) {
+            return Collections.emptyList();
+        }
+        getInstrumentation().logWarn("Bulk request failed");
+        handleBulkWriteErrors(bulkWriteErrors);
+
+        List<Message> failedMessages = new ArrayList<>();
+
+        bulkWriteErrors.forEach((bulkWriteError) -> {
```

Comment:
`return bulkWriteErrors.stream().filter(err -> !mongoRetryStatusCodeBlacklist.contains(String.valueOf(bulkWriteError.getCode())).collect(Collections.toList());`

Reply:
this will actually return the List <BulkWriteErrors >
we  need to return the failed List<Messages>

Reply:
return bulkWriteErrors.stream().filter(err -> !mongoRetryStatusCodeBlacklist.contains(String.valueOf(bulkWriteError.getCode())).map(err -> messages.get(err.getIndex()).collect(Collections.toList())

This one?

Reply:
yeah working now, thanks

Reply:
fixed